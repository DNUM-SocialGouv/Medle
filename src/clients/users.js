import fetch from "isomorphic-unfetch"

import { API_URL, RESET_PWD_ENDPOINT, USERS_ENDPOINT } from "../config"
import { handleAPIResponse } from "../utils/errors"
import { METHOD_DELETE, METHOD_PATCH, METHOD_POST, METHOD_PUT } from "../utils/http"

export const findUser = async ({ id, headers }) => {
  const response = await fetch(API_URL + USERS_ENDPOINT + "/" + id, { headers })
  return handleAPIResponse(response)
}

export const searchUsersFuzzy = async ({ search, requestedPage, headers }) => {
  const arr = []
  if (search) {
    arr.push(`fuzzy=${search}`)
  }
  if (requestedPage) {
    arr.push(`requestedPage=${requestedPage}`)
  }
  const bonus = arr.length ? "?" + arr.join("&") : ""
  const response = await fetch(`${API_URL}${USERS_ENDPOINT}${bonus}`, { headers })

  return handleAPIResponse(response)
}
export const createUser = async ({ user, headers }) => {
  const response = await fetch(`${API_URL}${USERS_ENDPOINT}`, {
    method: METHOD_POST,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(user),
  })
  return handleAPIResponse(response)
}

export const updateUser = async ({ user, headers }) => {
  const response = await fetch(`${API_URL}${USERS_ENDPOINT}/${user.id}`, {
    method: METHOD_PUT,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(user),
  })
  return handleAPIResponse(response)
}

export const deleteUser = async ({ id, headers }) => {
  const response = await fetch(`${API_URL}${USERS_ENDPOINT}/${id}`, { method: METHOD_DELETE, headers })
  return handleAPIResponse(response)
}

export const patchUser = async ({ id, password, headers }) => {
  const response = await fetch(`${API_URL}${RESET_PWD_ENDPOINT}`, {
    method: METHOD_PATCH,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify({ id, password }),
  })
  return handleAPIResponse(response)
}
