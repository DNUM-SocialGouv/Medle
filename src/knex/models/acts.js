import moment from "moment"

export const buildActsFromDB = acts =>
   acts.map(elt => ({
      id: elt.id,
      createdAt: elt.created_at,
      updatedAt: elt.updated_at,
      deletedAt: elt.deleted_at,
      casesFk: elt.cases_fk,
      pvNumber: elt.pv_number,
      examinationDate: elt.examination_date,
      asker: elt.asker,
      addedBy: elt.added_by,
      data: elt.data,
   }))

// from js to db
export const buildActFromJSON = data => {
   const res = { extra_data: {} }

   const mainKeys = {
      id: "id",
      internalNumber: "internal_number",
      pvNumber: "pv_number",
      examinationDate: "examination_date",
      profile: "profile",
      askerId: "asker_id",
      addedBy: "added_by",
      hospitalId: "hospital_id",
   }

   Object.keys(data).forEach(key => {
      if (mainKeys[key]) {
         res[mainKeys[key]] = data[key]
      } else {
         res.extra_data[key] = data[key]
      }
   })

   return res
}

// from db to js
export const buildActFromDB = data => ({
   id: data.id,
   internalNumber: data.internal_number,
   pvNumber: data.pv_number,
   examinationDate: moment(data.examination_date).format("YYYY-MM-DD"),
   profile: data.profile,
   asker: { id: data.asker_id, name: data.asker_name },
   user: {
      id: data.added_by,
      firstName: data.user_first_name,
      lastName: data.user_last_name,
      email: data.user_email,
   },
   hospital: { id: data.hospital_id, name: data.hospital_name },
   ...data.extra_data,
})
