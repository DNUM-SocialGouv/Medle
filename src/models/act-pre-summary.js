import moment from "moment"

import { ISO_DATE } from "../utils/date"

// from Knex entity to model (JS) entity
export const transform = (knexData) => {
    return !knexData
        ? null
        : {
            examinationDate: moment(knexData.examination_date).format(ISO_DATE),
            hospital: { id: knexData.hospital_id, name: knexData.hospital_name },
            examined: knexData.examined,
            actType: knexData.actType,
            violenceType: knexData.violenceType,
            location: knexData.location,
            age: knexData.age,
            actDuration: knexData.actDuration,
            category: knexData.category,
            ponderation: knexData.ponderation,
            id: knexData.id,
            user: {
                email: knexData.user_email,
                firstName: knexData.user_first_name,
                id: knexData.added_by,
                lastName: knexData.user_last_name,
            },
        }
}

export const transformAll = (list) => list.map((knexData) => transform(knexData))

// from model (JS) entity to Knex entity
export const untransform = (model) => {
    const knexData = { extra_data: {} }

    const mainKeys = {
        id: "id",
        examinationDate: "examination_date",
        category: "category",
        examined: "examined",
        actType: "act_type",
        violenceType: "violence_type",
        location: "location",
        age: "age",
        actDuration: "act_duration",
        ponderation: "ponderation",
        hospitalId: "hospital_id",
        addedBy: "added_by",
    }

    Object.keys(model).forEach((key) => {
        if (mainKeys[key]) {
            knexData[mainKeys[key]] = model[key]
        } else {
            knexData.extra_data[key] = model[key]
        }
    })

    return knexData
}
