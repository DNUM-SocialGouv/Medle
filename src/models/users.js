import * as yup from "yup"
import { logError } from "../utils/logger"
import { APIError } from "../utils/errors"
import { STATUS_400_BAD_REQUEST } from "../utils/http"

// from Knex entity to model (JS) entity
export const transform = (knexData) => {
  return !knexData
    ? null
    : {
        id: knexData.id,
        firstName: knexData.first_name,
        lastName: knexData.last_name,
        email: knexData.email,
        role: knexData.role,
        scope: knexData.scope,
        hospital: !knexData.hospital_id
          ? null
          : {
              id: knexData.hospital_id,
              name: knexData.hospital_name || "",
              ...knexData.hospital_extra_data,
            },
      }
}

export const transformAll = (list) => list.map((model) => transform(model))

// from model (JS) entity to Knex entity
export const untransform = (model) => {
  const knexData = {
    first_name: model.firstName,
    last_name: model.lastName,
    email: model.email,
    role: model.role,
    scope: !model.scope?.length ? null : JSON.stringify(model.scope.map((curr) => parseInt(curr.id, 10))), // Array needs to be explicitly stringified in PG
    hospital_id: (model.hospital && model.hospital.id) || null,
  }

  // Pas d'id pour une crÃ©ation de user
  if (model.id) knexData.id = model.id
  if (model.password) knexData.password = model.password

  return knexData
}

const schema = yup.object().shape({
  id: yup.number().positive().integer().nullable(),
  firstName: yup.string(),
  lastName: yup.string(),
  email: yup.string().required(),
  role: yup.string().required(),
  password: yup.string(),
  scope: yup
    .array()
    .default(() => [])
    .nullable(),

  hospital: yup
    .object()
    .shape({
      id: yup.number().positive().integer(),
      name: yup.string(),
    })
    .nullable(),
})

const configValidate = {
  strict: false,
  abortEarly: false,
}

export const validate = async (model) => {
  try {
    const value = await schema.validate(model, configValidate)
    return value
  } catch (error) {
    logError(error)
    throw new APIError({
      status: STATUS_400_BAD_REQUEST,
      message: "Bad request",
      detail: error.inner && error.inner.map((err) => [err.path, err.message]),
    })
  }
}

export const cast = async (model) => await schema.cast(model)
