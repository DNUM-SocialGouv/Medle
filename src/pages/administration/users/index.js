import React, { useState } from "react"
import Link from "next/link"
import { PropTypes } from "prop-types"
import { Alert, Col, Form, FormGroup, Input, Spinner, Table, Container } from "reactstrap"
import Layout from "../../../components/LayoutAdmin"
import fetch from "isomorphic-unfetch"

import { Title1 } from "../../../components/StyledComponents"
import { buildAuthHeaders, redirectIfUnauthorized, withAuthentication } from "../../../utils/auth"
import { API_URL, ADMIN_USERS_ENDPOINT } from "../../../config"
import { handleAPIResponse } from "../../../utils/errors"
import { logError } from "../../../utils/logger"
import { ADMIN, ROLES_DESCRIPTION } from "../../../utils/roles"
import { usePaginatedData } from "../../../utils/hooks"
import Pagination from "../../../components/Pagination"
import EditAttributesIcon from "@material-ui/icons/EditAttributes"
import AddIcon from "@material-ui/icons/Add"
import DoubleArrowIcon from "@material-ui/icons/DoubleArrow"
import { SearchButton } from "../../../components/form/SearchButton"

const fetchData = async ({ search, requestedPage, authHeaders }) => {
   const arr = []
   if (search) {
      arr.push(`fuzzy=${search}`)
   }
   if (requestedPage) {
      arr.push(`requestedPage=${requestedPage}`)
   }
   const bonus = arr.length ? "?" + arr.join("&") : ""
   const response = await fetch(`${API_URL}${ADMIN_USERS_ENDPOINT}${bonus}`, { headers: authHeaders })

   return handleAPIResponse(response)
}

const AdminUserPage = ({ paginatedData: initialPaginatedData, currentUser }) => {
   const [search, setSearch] = useState("")
   const [paginatedData, error, loading, fetchPage] = usePaginatedData(fetchData, initialPaginatedData)

   const onChange = e => {
      setSearch(e.target.value)
   }

   const onSubmit = e => {
      e.preventDefault()
      fetchPage(search)(0)
   }

   return (
      <Layout currentUser={currentUser}>
         <Container
            style={{ maxWidth: 980, minWidth: 740 }}
            className="mt-5 mb-4 d-flex justify-content-between align-items-center"
         >
            <Title1 className="">{"Administration des utilisateurs"}</Title1>
            <Link href="/administration/users/[id]" as={`/administration/users/new`}>
               <a>
                  <SearchButton className="mb-4 btn-outline-primary">
                     <AddIcon />
                     &nbsp; Ajouter
                  </SearchButton>
               </a>
            </Link>
         </Container>

         <Container style={{ maxWidth: 980, minWidth: 740 }}>
            <Form onSubmit={onSubmit}>
               <FormGroup row inline className="mb-4 justify-content-center">
                  <Col className="flex-grow-1">
                     <Input
                        type="text"
                        name="es"
                        id="es"
                        placeholder="Rechercher un utilisateur par nom, prénom, email.."
                        value={search}
                        onChange={onChange}
                        autoComplete="off"
                     />
                  </Col>
                  <Col className="flex-grow-0">
                     <SearchButton className="w-lg-75 btn-primary" disabled={loading}>
                        {loading ? <Spinner size="sm" color="light" data-testid="loading" /> : "Chercher"}
                     </SearchButton>
                  </Col>
               </FormGroup>
            </Form>
            {error && (
               <Alert color="danger" className="mb-4">
                  {error}
               </Alert>
            )}
            {!error && !paginatedData.elements.length && (
               <div className="text-center">{"Aucun utilisateurs trouvés."}</div>
            )}

            {!error && !!paginatedData.elements.length && (
               <>
                  <Pagination data={paginatedData} fn={fetchPage(search)} />
                  <Table responsive className="table-hover">
                     <thead>
                        <tr className="table-light">
                           <th>Nom</th>
                           <th>Email</th>
                           <th>Role</th>
                           <th>Établissement</th>
                           <th></th>
                           <th></th>
                        </tr>
                     </thead>
                     <tbody>
                        {paginatedData.elements.map(user => (
                           <tr key={user.id}>
                              <td>
                                 <b>{`${user.firstName} ${user.lastName}`}</b>
                              </td>
                              <td>{user.email}</td>
                              <td>{user.role && ROLES_DESCRIPTION[user.role]}</td>
                              <td>{user.hospital && user.hospital.name}</td>
                              <td>
                                 <Link href="/administration/users/[id]" as={`/administration/users/${user.id}`}>
                                    <a>
                                       Détails&nbsp;
                                       <EditAttributesIcon />
                                    </a>
                                 </Link>
                              </td>
                              <td>
                                 <Link
                                    href="/administration/users/reset/[id]"
                                    as={`/administration/users/reset/${user.id}`}
                                 >
                                    <a>
                                       Réinitialiser&nbsp;
                                       <DoubleArrowIcon />
                                    </a>
                                 </Link>
                              </td>
                           </tr>
                        ))}
                     </tbody>
                  </Table>
               </>
            )}
         </Container>
      </Layout>
   )
}

AdminUserPage.getInitialProps = async ctx => {
   const authHeaders = buildAuthHeaders(ctx)

   try {
      const paginatedData = await fetchData({ authHeaders })
      return { paginatedData }
   } catch (error) {
      logError("APP error", error)

      redirectIfUnauthorized(error, ctx)
   }
   return {}
}

AdminUserPage.propTypes = {
   paginatedData: PropTypes.object,
   currentUser: PropTypes.object.isRequired,
}

export default withAuthentication(AdminUserPage, ADMIN)
