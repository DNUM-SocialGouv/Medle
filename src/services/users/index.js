import { del } from "./delete"
import { find } from "./find"
import { search } from "./search"
import { create } from "./create"
import { update } from "./update"

export { create, del, find, search, update }
