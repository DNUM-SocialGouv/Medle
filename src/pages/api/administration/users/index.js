import Cors from "micro-cors"

import { STATUS_200_OK, METHOD_GET, METHOD_OPTIONS, METHOD_POST } from "../../../../utils/http"
import { sendAPIError, sendMethodNotAllowedError } from "../../../../services/errorHelpers"
import { checkValidUserWithPrivilege } from "../../../../utils/auth"
import { ADMIN } from "../../../../utils/roles"

import { search, create } from "../../../../services/users"

const handler = async (req, res) => {
   res.setHeader("Content-Type", "application/json")

   try {
      switch (req.method) {
         case METHOD_GET: {
            const currentUser = checkValidUserWithPrivilege(ADMIN, req, res)

            const { users, totalCount, currentPage, maxPage, byPage } = await search({ ...req.query, currentUser })

            return res.status(STATUS_200_OK).json({ totalCount, currentPage, maxPage, byPage, elements: users })
         }
         case METHOD_POST: {
            const currentUser = checkValidUserWithPrivilege(ADMIN, req, res)

            const id = await create(req.body, currentUser)

            return res.status(STATUS_200_OK).json({ id })
         }

         default:
            return sendMethodNotAllowedError(res)
      }
   } catch (error) {
      sendAPIError(error, res)
   }
}

const cors = Cors({
   allowMethods: [METHOD_GET, METHOD_OPTIONS, METHOD_POST],
})

export default cors(handler)
