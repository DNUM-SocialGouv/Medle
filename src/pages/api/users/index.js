import Cors from "micro-cors"

import { sendAPIError, sendMethodNotAllowedError } from "../../../services/errorHelpers"
import { create, search } from "../../../services/users"
import { checkIsSuperAdmin, checkValidUserWithPrivilege } from "../../../utils/auth"
import { METHOD_GET, METHOD_OPTIONS, METHOD_POST, STATUS_200_OK } from "../../../utils/http"
import { ADMIN } from "../../../utils/roles"

const handler = async (req, res) => {
  res.setHeader("Content-Type", "application/json")

  try {
    switch (req.method) {
      case METHOD_GET: {
        const currentUser = checkValidUserWithPrivilege(ADMIN, req, res)

        checkIsSuperAdmin(currentUser)

        const { users, totalCount, currentPage, maxPage, byPage } = await search({ ...req.query, currentUser })

        return res.status(STATUS_200_OK).json({ totalCount, currentPage, maxPage, byPage, elements: users })
      }
      case METHOD_POST: {
        const currentUser = checkValidUserWithPrivilege(ADMIN, req, res)

        checkIsSuperAdmin(currentUser)

        const id = await create(req.body, currentUser)

        return res.status(STATUS_200_OK).json({ id })
      }
      default:
        if (req.method !== METHOD_OPTIONS) return sendMethodNotAllowedError(res)
    }
  } catch (error) {
    sendAPIError(error, res)
  }
}

const cors = Cors({
  allowMethods: [METHOD_GET, METHOD_OPTIONS, METHOD_POST],
})

export default cors(handler)
