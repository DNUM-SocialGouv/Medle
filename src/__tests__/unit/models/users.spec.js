import { transform, untransform } from "../../../models/users"

describe("transform", () => {
  it("should transform in a model entity", () => {
    const knexUser = {
      id: 2,
      first_name: "John",
      last_name: "Mc Lane",
      email: "john.mclane@proton.mail",
      role: "OPERATOR_ACT",
      hospital_id: 3,
    }

    expect(transform(knexUser)).toMatchSnapshot()
  })
})

describe("untransform", () => {
  it("should untransform in a Knex entity with an id and without a password", () => {
    const modelUser = {
      id: 1,
      firstName: "Marty",
      lastName: "Mac Fly",
      email: "marty.macfly@gmail.com",
      role: "REGIONAL_SUPERVISOR",
      scope: [{ id: 1 }, { id: 3 }, { id: 6 }],
      hospital: null,
    }

    expect(untransform(modelUser)).toMatchSnapshot()
  })
  it("should untransform in a Knex entity even without id and wih a password", () => {
    const modelUser = {
      id: 2,
      firstName: "Doc",
      lastName: "Brown",
      email: "doc.brown@yahoo.com",
      password: "12345678",
      role: "OPERATOR_ACT",
      hospital: 2,
    }

    expect(untransform(modelUser)).toMatchSnapshot()
  })
  it("should untransform in a Knex entity even without id and password", () => {
    const modelUser = {
      firstName: "Doc",
      lastName: "Brown",
      email: "doc.brown@yahoo.com",
      role: "OPERATOR_ACT",
      hospital: 2,
    }

    expect(untransform(modelUser)).toMatchSnapshot()
  })
})
