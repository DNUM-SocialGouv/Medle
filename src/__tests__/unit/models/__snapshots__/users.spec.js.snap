// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform should transform in a model entity 1`] = `
Object {
  "email": "john.mclane@proton.mail",
  "firstName": "John",
  "hospital": Object {
    "id": 3,
    "name": "",
  },
  "id": 2,
  "lastName": "Mc Lane",
  "role": "OPERATOR_ACT",
  "scope": undefined,
}
`;

exports[`untransform should untransform in a Knex entity even without id and password 1`] = `
Object {
  "email": "doc.brown@yahoo.com",
  "first_name": "Doc",
  "hospital_id": null,
  "last_name": "Brown",
  "role": "OPERATOR_ACT",
  "scope": null,
}
`;

exports[`untransform should untransform in a Knex entity even without id and wih a password 1`] = `
Object {
  "email": "doc.brown@yahoo.com",
  "first_name": "Doc",
  "hospital_id": null,
  "id": 2,
  "last_name": "Brown",
  "password": "12345678",
  "role": "OPERATOR_ACT",
  "scope": null,
}
`;

exports[`untransform should untransform in a Knex entity with an id and without a password 1`] = `
Object {
  "email": "marty.macfly@gmail.com",
  "first_name": "Marty",
  "hospital_id": null,
  "id": 1,
  "last_name": "Mac Fly",
  "role": "REGIONAL_SUPERVISOR",
  "scope": "[1,3,6]",
}
`;
